### PR TITLE
filterValues and filterKeys

### PR DESCRIPTION
**Pull request by matthewleon**
supersedes https://github.com/purescript/purescript-maps/pull/80

Rebases and implements the suggestions in that PR.

### CODE REVIEW

File: src/Data/StrMap.purs [hunk 1]
```diff
@@ -259,3 +262,26 @@ instance semigroupStrMap :: (Semigroup a) => Semigroup (StrMap a) where
 
 instance monoidStrMap :: (Semigroup a) => Monoid (StrMap a) where
   mempty = empty
+
+-- | Filter out those key/value pairs of a map for which a predicate
+-- | fails to hold.
+filter :: forall a. (String -> a -> Boolean) -> StrMap a -> StrMap a
```
Comment (paf31) [inline src/Data/StrMap.purs]: `filterWithKey` perhaps?

File: src/Data/StrMap.purs [hunk 2]
```diff
@@ -259,3 +262,26 @@ instance semigroupStrMap :: (Semigroup a) => Semigroup (StrMap a) where
...
+    step acc k v = if predicate k v then SM.poke acc k v else pure acc
+
+-- | Filter out those key/value pairs of a map for which a predicate
+-- | on the key fails to hold.
+filterKeys :: (String -> Boolean) -> StrMap ~> StrMap
+filterKeys predicate = filter $ const <<< predicate
+
+-- | Filter out those key/value pairs of a map for which a predicate
+-- | on the value fails to hold.
+filterValues :: forall a. (a -> Boolean) -> StrMap a -> StrMap a
```
Comment (paf31) [inline src/Data/StrMap.purs]: I would just call this one `filter` actually.

**Comment by paf31** [comment]
Sorry for the bikeshedding 😁 

cc @hdgarrood

**Comment by matthewleon** [comment]
> Sorry for the bikeshedding 😁

Nothing to be sorry about, someone's gotta do it!

**Comment by hdgarrood** [comment]
:+1: for `filter` and `filterWithKey`.

**Comment by paf31** [comment]
Thanks!

**Pull request merged**